### PR TITLE
Version number simplification for MV3

### PR DESCRIPTION
**Pull request by rebloor**
### Description

Updates the information about version number format to provide for the simplified format for MV3 extensions in Firefox 108 and later.

### Related issues and pull requests

Addresses the documentation needs of [Bug 1793925](https://bugzilla.mozilla.org/show_bug.cgi?id=1793925) Warn about complex versioning formats used in manifest

### CODE REVIEW

File: files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md [hunk 1]
```diff
@@ -9,41 +9,55 @@ tags:
...
 - if string-b is a plus sign, number-a is incremented to be compatible with the Firefox 1.0.x version format:
   `1.0+` is the same as `1.1pre`
 
-The rationale behind splitting a version part into a sequence of strings and numbers is that when comparing version parts, the numeric parts are compared as numbers, e.g. '1.0pre1' < '1.0pre10', while the strings are compared bytewise. See the next section for details on how versions are compared.
+The rationale behind splitting a version part into a sequence of strings and numbers is that when comparing version parts, the numeric parts are compared as numbers, for example, '1.0pre1' < '1.0pre10', while the strings are compared byte-wise. See the next section for details on how versions are compared.
 
-See [maintenance policy](https://extensionworkshop.com/documentation/publish/signing-and-distribution-overview/#distributing-your-addon) for special rules that mark version as "beta" in the AMO queues.
+See [maintenance policy](https://extensionworkshop.com/documentation/publish/signing-and-distribution-overview/#distributing-your-addon) for special rules that mark a version as "beta" in the AMO queues.
```
Comment (rebloor) [inline files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md:44]: @Rob--W @willdurand any thoughts on what this refers to? There's no obvious information on the link page or  related pages about beta releases that I can see. Haven't had time to dig through AMO yet though.

File: files/en-us/mozilla/add-ons/webextensions/manifest.json/version/index.md [hunk 2]
```diff
@@ -27,7 +27,7 @@ browser-compat: webextensions.manifest.version
   </tbody>
 </table>
 
-Version of the extension, formatted as numbers and ASCII characters separated by dots. For the details of the version format, see the [Version format](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/version/format) page.
+Version of the extension, formatted as numbers and ASCII characters or, for Manifest V3 staring in Firefox 108, only numbers separated by dots in both cases. See [Manifest Version Format](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/version/format), for details about the version format.
 
 You can inspect the [add-ons linter code](https://github.com/mozilla/addons-linter/blob/master/src/schema/formats.js#L10) to see how extension versions for Firefox are validated.
```
Comment (rebloor) [inline files/en-us/mozilla/add-ons/webextensions/manifest.json/version/index.md:32]: @willdurand  @Rob--W can you? I don't see any obvious sign of this including the MV3 format yet

File: files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md (hunk 1 above)
Comment (rebloor) [reply files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md:44]: So, unless something has changed and we haven't updated extension workshop documentation, I don't believe we have the notion of beta in the "AMO queues.." You can flag an extension as experimental or, if you want to run a beta, distribute it yourself (https://extensionworkshop.com/documentation/publish/distribute-pre-release-versions/). So I'm thinking we should remove this reference. Agree?

Comment (willdurand) [reply files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md:44]: I think you're right, beta is no longera thing on AMO (removed in 2019 iirc)

File: files/en-us/mozilla/add-ons/webextensions/manifest.json/version/index.md (hunk 2 above)
Comment (willdurand) [reply files/en-us/mozilla/add-ons/webextensions/manifest.json/version/index.md:32]: In 108, Firefox emits warnings when the version string does not match the new format.

Comment (rebloor) [reply files/en-us/mozilla/add-ons/webextensions/manifest.json/version/index.md:32]: @willdurand so is this more of a recommended version number structure than a required one? Will AMO continue to accept MV2-style version numbers for MV3 extensions?

Comment (willdurand) [reply files/en-us/mozilla/add-ons/webextensions/manifest.json/version/index.md:32]: AMO will continue to accept MV2-style version numbers for MV2 extensions (for now). For MV3 extensions, only the new format will be allowed.

File: files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md [hunk 3]
```diff
@@ -9,41 +9,53 @@ tags:
...
 
-A few examples of valid version parts:
+From Firefox 108, the **version string** must be 1 to 4 numbers separated by dots, for example, `1.2.3.4`. Each number can have up to 9 digits. However, a leading zero before another digit is not allowed. For example, `2.01` is not an allowed version number; however, `0.2`, `2.0.1`, and `2.10` are allowed.
+
+The regular expression for this version string is `^(0|[1-9]\d{0,9})(\.(0|[1-9]\d{0,9})){0,3}$`.
+
+This **version string** is compatible with Manifest V3 in Firefox 107 or earlier and Manifest V2.
+
+## Version numbers in Manifest V2
+
```
Comment (willdurand) [inline files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md:23]: I'd add a note to encourage developers to switch to the newer version string format because AMO will soon emit warnings and we might want to be more strict in the future.

File: files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md [hunk 4]
```diff
@@ -9,41 +9,53 @@ tags:
 
 {{AddonSidebar}}
 
-A **version string** consists of one or more _version parts_, separated with dots.
+## Version numbers in Manifest V3
```
Comment (willdurand) [inline files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md:12]: It will be enforced in MV3 but I think the title should be "recommended version numbers" or something like that.

File: files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md [hunk 5]
```diff
@@ -9,41 +9,53 @@ tags:
...
+The **version string** for Manifest V3 extensions was simplified in Firefox 108.
 
-A few examples of valid version parts:
+From Firefox 108, the **version string** must be 1 to 4 numbers separated by dots, for example, `1.2.3.4`. Each number can have up to 9 digits. However, a leading zero before another digit is not allowed. For example, `2.01` is not an allowed version number; however, `0.2`, `2.0.1`, and `2.10` are allowed.
+
+The regular expression for this version string is `^(0|[1-9]\d{0,9})(\.(0|[1-9]\d{0,9})){0,3}$`.
+
+This **version string** is compatible with Manifest V3 in Firefox 107 or earlier and Manifest V2.
+
+## Version numbers in Manifest V2
```
Comment (willdurand) [inline files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md:22]: Echo'ing the comment above, I'd rename this title to "legacy version numbers (Manifest V2)" maybe.

File: files/en-us/mozilla/add-ons/webextensions/manifest.json/version/index.md [hunk 6]
```diff
@@ -27,7 +27,7 @@ browser-compat: webextensions.manifest.version
   </tbody>
 </table>
 
-Version of the extension, formatted as numbers and ASCII characters separated by dots. For the details of the version format, see the [Version format](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/version/format) page.
+Version of the extension, formatted as numbers and ASCII characters or, for Manifest V3 staring in Firefox 108, only numbers separated by dots in both cases. See [Manifest Version Format](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/version/format), for details about the version format.
```
Comment (willdurand) [inline files/en-us/mozilla/add-ons/webextensions/manifest.json/version/index.md:30]: It'd be good to only mention the newer format given it is compatible with the legacy one. We should no longer encourage the use of letters in the version string, even though it is still accepted currently.

File: files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md (hunk 4 above)
Comment (rebloor) [reply files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md:12]: Using the term "recommended" could suggest to developers that they can continue using the "old" version numbers, however, if they're enforced for MV3 that doesn't seem like the case. I'd suggest keeping the title as it is.

File: files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md (hunk 5 above)
Comment (Rob--W) [reply files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md:22]: I second this request. Don't call it "manifest v2" version format, but legacy (or "toolkit version format"). Refer to the new one as the "extension version format". Anything valid in the "new" version format is also valid in the "old" format, i.e. the new format is a subset of the legacy one.

To avoid confusion, it may be worth having a section that explains the different "version" meanings, so that if a developer ends up on the wrong page, that they can easily jump to the right one. Completely unrelated, but all "version":
- manifest_version 2 and 3 atm
- version - extension version (format subject of this PR)
- browser_specific_settings.gecko.strict_min_version (and also strict_max_version) - Firefox version. May contain the letters "a" or "b". Examples are 100.0, 100.0b7, 100.0a1 for Release, Beta and Nightly. Release may also look like 100.0.1. The second digit may also be non-zero for Firefox ESR.

Comment (rebloor) [reply files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md]: @willdurand @Rob--W apologies, I updated this yesterday but forgot to update the review request. I have:

- revised the Version page to make it generic with the links to further information for Firefox (this format page) and an external link for Chrome. I couldn't find anything for Safari and still need to check whether there's anything specific for Edge
- move the Firefox-specific information from the Version page to here
- follow your suggested naming for the version formats/page sections

File: files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md [hunk 7]
```diff
@@ -9,41 +9,62 @@ tags:
...
   `1.5.0.*` is the same as `1.5.0.(infinity)`
 - if string-b is a plus sign, number-a is incremented to be compatible with the Firefox 1.0.x version format:
   `1.0+` is the same as `1.1pre`
 
-The rationale behind splitting a version part into a sequence of strings and numbers is that when comparing version parts, the numeric parts are compared as numbers, e.g. '1.0pre1' < '1.0pre10', while the strings are compared bytewise. See the next section for details on how versions are compared.
+The rationale behind splitting a version part into a sequence of strings and numbers is that when comparing version parts, the numeric parts are compared as numbers, for example, '1.0pre1' < '1.0pre10', while the strings are compared byte-wise. See the next section for details on how versions are compared.
 
-See [maintenance policy](https://extensionworkshop.com/documentation/publish/signing-and-distribution-overview/#distributing-your-addon) for special rules that mark version as "beta" in the AMO queues.
+You can inspect the [add-ons linter code](https://github.com/mozilla/addons-linter/blob/master/src/schema/formats.js#L10) to see how extension versions for Firefox are validated.
```
Comment (Rob--W) [inline files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md:53]: Let's remove this. The description in the docs here should be self-sufficient.

And if you ever want to link to github (or gitlab or any other code hosting site), try to use a permalink. A link to a (master) branch can and will eventually become inaccurate when the source code updates. Reliable permalinks are the commit hash or a (release) tag. To get the permalink on github, press the Y shortcut, for example. For documentation, see https://docs.github.com/en/repositories/working-with-files/using-files/getting-permanent-links-to-files

File: files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md [hunk 8]
```diff
@@ -9,41 +9,62 @@ tags:
...
 
-## Comparing versions
+### Comparing versions
 
-When two version strings are compared, their version parts are compared left to right. An empty or missing version part is equivalent to `0`.
+When two version strings are compared, their version parts are compared left to right. For Manifest V2 style version strings, an empty or missing version part is equivalent to `0`.
```
Comment (Rob--W) [inline files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md:57]: Isn't this description accurate for all version formats? 1.0 is the same as 1.0.0 for example. I'm trying to understand why the qualifier "MV2-only" was applied here.

(And if it had to be kept, it should probably have been referred to as legacy, not MV2)

File: files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md [hunk 9]
```diff
@@ -9,41 +9,62 @@ tags:
...
+Otherwise, the version strings are equal. Note that as missing version parts are treated as if they were `0`, these version strings are equal: `1`, `1.0`, `1.0.`, `1.0.0`, and even `1.0..`.
 
-### Comparing version parts
+#### Comparing version parts
 
-Version parts are also compared left to right, parts A and C are compared as numbers, while parts B and D are compared byte-wise. A string-part that exists is always less than a string-part that doesn't exist (`1.6a` is less than `1.6`).
+Version parts are also compared left to right, parts A and C are compared as numbers, while parts B and D are compared byte-wise. A string part that exists is always less than a string part that doesn't exist. For example, `1.6a` is less than `1.6`.
 
-## Examples
+#### Bxamples
```
Comment (Rob--W) [inline files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md:67]: Bxamples?

File: files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md (hunk 7 above)
Comment (rebloor) [reply files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md:53]: This was a copy-paste of the existing line - but thanks for the hint/reminder - removed.

File: files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md (hunk 8 above)
Comment (rebloor) [reply files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md:57]: I have reverted to the original "legacy" version number description and added the comment about zero equivalency to the extension version number comparison description

File: files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md (hunk 9 above)
Comment (rebloor) [reply files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md]: Fixed

File: files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md [hunk 10]
```diff
@@ -9,39 +9,41 @@ tags:
 
 {{AddonSidebar}}
```
Comment (Rob--W) [inline files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md:11]: Could you prepend an introduction to point visitors (arriving from search engines for example) to the correct places? E.g. something like this

> This page documents the version string format used by legacy Firefox extensions and WebExtensions up to Manifest Version 2. For documentation on the currently recommended version strings, see [link to the parent manifest/version article].
> For information about the manifest version (distinct from extension version), see [link to manifest_version documentation]

The changelog links to this `version/format` page, so it would be nice to have a quick link to the correct location.

File: files/en-us/mozilla/add-ons/webextensions/manifest.json/version/index.md [hunk 11]
```diff
@@ -27,34 +27,47 @@ browser-compat: webextensions.manifest.version
...
 
-Note that [the syntax defined for Chrome's `version`](https://developer.chrome.com/docs/extensions/mv3/manifest/version/) is more restrictive than that used by Firefox:
+The version string consists of 1 to 4 numbers separated by dots, for example, `1.2.3.4`. Non-zero numbers must not include a leading zero. For example, `2.01` is not allowed; however, `0.2`, `2.0.1`, and `2.10` are allowed.
 
-- values for `version` that are valid for Chrome will always be valid for Firefox
-- values for `version` that are valid for Firefox may not be valid for Chrome
+Extension stores and browsers may enforce or warn if the version string doesn't comply with this format. They may also apply restrictions to the range of numbers available. For example:
 
-To obtain your extension version in your JavaScript code, use:
+- Firefox Browser Add-ons requires version strings using numbers of up to nine digits, complying with this regular expression `^(0|[1-9]\d{0,9})(\.(0|[1-9]\d{0,9})){0,3}$`. Also, from Firefox 108, a warning is provided if an extension is installed with a version number that doesn't match this format.
```
Comment (Rob--W) [inline files/en-us/mozilla/add-ons/webextensions/manifest.json/version/index.md:38]:
```suggestion
- Firefox requires extension version strings using numbers of up to nine digits, complying with this regular expression `^(0|[1-9][0-9]{0,8})([.](0|[1-9][0-9]{0,8})){0,3}$`. Also, from Firefox 108, a warning is provided if an extension is installed with a version number that doesn't match this format.
```

Not sure what "Firefox Browser Add-ons" is referring to. Firefox browser or AMO? I changed it to Firefox, but if you meant AMO (addons.mozilla.org), please change it accordingly.

I fixed the RegExp as follows:
- Use `{0,8}` instead of `{0,9}`. Your pattern already has a `[1-9]` at the front, so `{0,9}` would actually make the total 10 digits.
- I replaced `\d` with `[0-9]`. `\d` is not wrong, but there is a lower chance of incorrect usage when the pattern does not contain a backslash. For example, if someone were to copy-paste this pattern in a string, then the backslash would be interpreted as a string escape instead of a regexp escape character.
- `\.` was replaced with `[.]` for the same reason.

**Review by Rob--W — COMMENTED** [review]
I really like the direction of this PR. The readability and level of information has improved significantly.

File: files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md (hunk 10 above)
Comment (rebloor) [reply files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md:11]: @Rob--W I have added an introductory paragraph, however, I've kept it generic. We also don't have a page that talks about manifest versions, so I haven't added the final part. Hopefully, as the page title no longer says "manifest version" there is less likely to be confusion. I have also simplified the release note, as I don't think this really needs to include the details of the extension version format and I've also changed it to point to the Version key page.

File: files/en-us/mozilla/add-ons/webextensions/manifest.json/version/index.md [hunk 12]
```diff
@@ -27,34 +27,47 @@ browser-compat: webextensions.manifest.version
...
+- The Chrome Web Store requires [numbers between 0 and 65535](https://developer.chrome.com/docs/extensions/mv3/manifest/version/) and does not permit all-zero extension strings. For example 0.0 or 0.0.0.0 are not permitted.
+
+It may be possible to create an extension that appears to have a valid version number when run in a browser but doesn't comply with store requirements. Particular care should be taken when developing cross-browser extensions that use large number elements.
+
+Use [version_name](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/version_name) if you want to include an identifier in the version number that isn't supported by the store or browser. For example, `1.0 beta`.
```
Comment (willdurand) [inline files/en-us/mozilla/add-ons/webextensions/manifest.json/version/index.md:43]: I don't think Firefox or AMO rely on that property.

**Review by willdurand — APPROVED** [review]
Once Rob's comments have been addressed, r+wc from my side

File: files/en-us/mozilla/add-ons/webextensions/manifest.json/version/index.md (hunk 12 above)
Comment (Rob--W) [reply files/en-us/mozilla/add-ons/webextensions/manifest.json/version/index.md:43]: @willdurand's remark is correct. Let's rephrase it to emphasize that its support is not universal. For example (please reword to make it flow better):

> Some browsers and web stores may recognize the [version_name](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/version_name) to complement the `version` key, with a more permissive string. For example, `1.0 beta`.

**Review by Rob--W — COMMENTED** [review]
The last update didn't cover https://github.com/mdn/content/pull/21699#discussion_r1008309117. Was that intentional or pending clarification?

Other than that and the comments here, it's basically ready to merge.

File: files/en-us/mozilla/add-ons/webextensions/manifest.json/version/index.md (hunk 11 above)
Comment (rebloor) [reply files/en-us/mozilla/add-ons/webextensions/manifest.json/version/index.md:38]: Thanks - updated to AMO and updated the regexp.

File: files/en-us/mozilla/add-ons/webextensions/manifest.json/version/index.md (hunk 12 above)
Comment (rebloor) [reply files/en-us/mozilla/add-ons/webextensions/manifest.json/version/index.md:43]: Correct - reworded

File: files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md (hunk 10 above)
Comment (rebloor) [reply files/en-us/mozilla/add-ons/webextensions/manifest.json/version/format/index.md:11]: @Rob--W not sure if you were asking for this to be added. Given the page changes I'm not sure there is a significant possibility of this page getting confused with manifest versions.

**Comment by rebloor** [comment]
@Rob--W I believe I've addressed all your feedback. I'll merge tomorrow if you have no further feedback.

**Review by Rob--W — APPROVED** [review]
Thanks!
